src/tests/ftest/util/rebuild_test_base.py

```diff
@@ -0,0 +1,222 @@
+#!/usr/bin/python
+"""
+  (C) Copyright 2018-2019 Intel Corporation.
+
+  Licensed under the Apache License, Version 2.0 (the "License");
+  you may not use this file except in compliance with the License.
+  You may obtain a copy of the License at
+
+     http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+
+  GOVERNMENT LICENSE RIGHTS-OPEN SOURCE SOFTWARE
+  The Government's rights to use, modify, reproduce, release, perform, display,
+  or disclose this software are subject to the terms of the Apache License as
+  provided in Contract No. B609815.
+  Any reproduction of computer software, computer software documentation, or
+  portions thereof marked with this legend must also reproduce the markings.
+"""
+from apricot import TestWithServers
+from command_utils import ObjectWithParameters, BasicParameter
+from test_utils import TestPool, TestContainer
+
+
+class RebuildTestParams(ObjectWithParameters):
+    """Class for gathering test parameters."""
+
+    def __init__(self):
+        """Initialize a RebuildTestParams object."""
+        self.object_class = BasicParameter(None)
+        self.rank = BasicParameter(None)
+
+
+class RebuldTestBase(TestWithServers):
+    """Base rebuild test class.
+
+    :avocado: recursive
+    """
+
+    def __init__(self, *args, **kwargs):
+        """Initialize a RebuldTestBase object."""
+        super(RebuldTestBase, self).__init__(*args, **kwargs)
+        self.inputs = RebuildTestParams()
+        self.targets = None
+        self.container = None
+        self.server_count = 0
+        self.info_checks = None
+        self.rebuild_checks = None
+
+    def setUp(self):
+        """Set up each test case."""
+        # Start the servers and agents
+        super(RebuldTestBase, self).setUp()
+
+        # Get the test parameters
+        self.inputs.get_params(self, "/run/test/*")
+
+        # Get the number of targets per server for pool info calculations
+        self.targets = self.params.get("targets", "/run/server_config/*")
+
+        self.server_count = len(self.hostlist_servers)
+
+    def tearDown(self):
+        """Tear down each test case."""
+        try:
+            # Destroy the containers - assuming a TestContainer object
+            if self.container is not None:
+                self.container.destroy(1)
+
+            # Destroy the pools - assuming a TestPool object
+            if self.pool is not None:
+                self.pool.destroy(1)
+
```

Comment:
Not sure we want a blank line here, but much more importantly, we need to catch exceptions here and use `self.fail()` to indicate them to Avocado rather than just pushing the exception up to Avocado.  This is essentially what [DAOS-3203](https://jira.hpdd.intel.com/browse/DAOS-3203) is about and what Avocado [documents](https://avocado-framework.readthedocs.io/en/52.1/WritingTests.html#turning-errors-into-failures﻿​_).

And yes, the Functional tests are rife with this issue.

It would be great if we could set some kind of  "global" `fail_on()` for the `DaosApiError()` exception, but I know of no way to do that.

I also wonder, given the amount of pool-destroy-without-exception-handling that we probably have in our Functional tests if it's time to formalise it into _apricot_ in some way so that every test writer doesn't have to remember to do it properly.  Perhaps a `TestWithPool` subclass.  Or some other means.  Maybe it doesn't need a whole subclass.

Reply:
I agree on the tear down comments.  I'm waiting to land #1066 before proceeding with this pull request.

Reply:
#1066 has landed successfully, so I'm  marking this resolved.